src/lib/mixer/Mixer.hpp

```diff
@@ -0,0 +1,277 @@
+ */
+class Mixer : public ListNode<Mixer *>
+{
+public:
+	enum class Airmode : int32_t {
+		disabled = 0,
+		roll_pitch = 1,
+		roll_pitch_yaw = 2
+	};
+
+	/**
+	 * Fetch a control value.
+	 *
+	 * @param handle		Token passed when the callback is registered.
+	 * @param control_group		The group to fetch the control from.
+	 * @param control_index		The group-relative index to fetch the control from.
+	 * @param control		The returned control
+	 * @return			Zero if the value was fetched, nonzero otherwise.
+	 */
+	typedef int	(* ControlCallback)(uintptr_t handle, uint8_t control_group, uint8_t control_index, float &control);
+
+	/**
+	 * Constructor.
+	 *
+	 * @param control_cb		Callback invoked when reading controls.
+	 */
+	Mixer(ControlCallback control_cb, uintptr_t cb_handle) : _control_cb(control_cb), _cb_handle(cb_handle) {}
+	virtual ~Mixer() = default;
+
+	// no copy, assignment, move, move assignment
+	Mixer(const Mixer &) = delete;
+	Mixer &operator=(const Mixer &) = delete;
+	Mixer(Mixer &&) = delete;
+	Mixer &operator=(Mixer &&) = delete;
+
+	/**
+	 * Perform the mixing function.
+	 *
+	 * @param outputs		Array into which mixed output(s) should be placed.
+	 * @param space			The number of available entries in the output array;
+	 * @return			The number of entries in the output array that were populated.
+	 */
+	virtual unsigned		mix(float *outputs, unsigned space) = 0;
+
+	/**
+	 * Get the saturation status.
+	 *
+	 * @return			Integer bitmask containing saturation_status from multirotor_motor_limits.msg.
+	 */
+	virtual uint16_t		get_saturation_status() { return 0; }
+
+	/**
+	 * Analyses the mix configuration and updates a bitmask of groups
+	 * that are required.
+	 *
+	 * @param groups		A bitmask of groups (0-31) that the mixer requires.
+	 */
+	virtual void			groups_required(uint32_t &groups) {};
+
+	/**
+	 * @brief      Empty method, only implemented for MultirotorMixer and MixerGroup class.
+	 *
+	 * @param[in]  delta_out_max  Maximum delta output.
+	 *
+	 */
+	virtual void 			set_max_delta_out_once(float delta_out_max) {}
+
+	/**
+	 * @brief Set trim offset for this mixer
+	 *
+	 * @return the number of outputs this mixer feeds to
+	 */
+	virtual unsigned		set_trim(float trim) { return 0; }
+
+	/**
+	 * @brief Get trim offset for this mixer
+	 *
+	 * @return the number of outputs this mixer feeds to
+	 */
+	virtual unsigned		get_trim(float *trim) { return 0; }
+
+	/*
+	 * @brief      Sets the thrust factor used to calculate mapping from desired thrust to motor control signal output.
+	 *
+	 * @param[in]  val   The value
+	 */
+	virtual void 			set_thrust_factor(float val) {}
+
+	/**
+	 * @brief Set airmode. Airmode allows the mixer to increase the total thrust in order to unsaturate the motors.
+	 *
+	 * @param[in]  airmode   Select airmode type (0 = disabled, 1 = roll/pitch, 2 = roll/pitch/yaw)
+	 */
+	virtual void			set_airmode(Airmode airmode) {};
+
+	virtual unsigned		get_multirotor_count()  { return 0; }
```

Comment:
Hi @dagar, do you think this should be bumped to the MultiRotor Mixer? (It appears to me that the helicopter mixer is not utilizing it.)  Everything else is looking pretty good in this PR.

Reply:
No, unfortunately with the current structure any interaction we want with any mixer has to go through a common interface.